### PR TITLE
Replace RNG update in RV lift rewrites

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description

By only replacing the output variable (dimshuffle, subtensor), we were leaving the second output of the lifted RV (the rng update) linked to the old node. In a normal function this RNG update is an output of the function and we want to also replace it to link to the new RV node.

Not doing this would result in "invalid random graphs", with the same RNG leading to two RVs of different sizes/shapes.

Also simplified a bit the rewrites in the first commit, now that RandomVariable always adds expand dims to the batch dimensions of the parameters

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/tensor/random/rewriting/basic.py [hunk 1]
```diff
@@ -189,10 +183,17 @@ def local_dimshuffle_rv_lift(fgraph, node):
...
-        out.name = f"{base_rv.name}_lifted"
-    return [out]
+    new_next_rng, new_rv = new_node.outputs
+
+    if rv.name:
+        new_rv.name = f"{rv.name}_lifted"
+
+    # We replace uses of the dimshuffled RV by the new RV
+    # And uses of the old RNG update by the new RNG update
+    return {
```
Comment (lucianopaz) [inline pytensor/tensor/random/rewriting/basic.py:193]: I'm not familiar with this type of output. I've always seen lists as return types. What happens with dictionaries? Are they taken as a sort of `updates`?

File: pytensor/tensor/random/rewriting/basic.py [hunk 2]
```diff
@@ -153,31 +154,24 @@ def local_dimshuffle_rv_lift(fgraph, node):
...
-        # Make size explicit
-        shape = tuple(broadcast_params(dist_params, rv_op.ndims_params)[0].shape)
-        size = shape[:batched_dims]
-
-    # Update the size to reflect the DimShuffled dimensions
-    new_size = [
-        constant(1, dtype="int64") if o == "x" else size[o]
-        for o in batched_dims_ds_order
-    ]
+        new_size = size
```
Comment (lucianopaz) [inline pytensor/tensor/random/rewriting/basic.py:164]: Why were you able to remove all of the original code that was visited in this branch?

File: pytensor/tensor/random/rewriting/basic.py [hunk 3]
```diff
@@ -206,47 +207,38 @@ def local_subtensor_rv_lift(fgraph, node):
...
-
-    # Use shape_feature to facilitate inferring final shape.
-    # Check that neither the RV nor the old Subtensor are in the shape graph.
-    output_shape = fgraph.shape_feature.shape_of.get(old_subtensor, None)
-    if output_shape is None or {old_subtensor, rv} & set(ancestors(output_shape)):
-        return None
-
     rv_op = rv_node.op
     rng, size, *dist_params = rv_node.inputs
+    next_rng, rv = rv_node.outputs
```
Comment (lucianopaz) [inline pytensor/tensor/random/rewriting/basic.py:228]: Why could you remove the `shape_feature` stuff?

File: pytensor/tensor/random/rewriting/basic.py [hunk 4]
```diff
@@ -277,34 +269,35 @@ def is_nd_advanced_idx(idx, dtype):
...
-        return False
-
     # Update the size to reflect the indexed dimensions
-    new_size = output_shape[: len(output_shape) - rv_op.ndim_supp]
+    if isinstance(size.type, NoneTypeT):
+        new_size = size
+    else:
+        shape_feature = getattr(fgraph, "shape_feature", None)
+        if not shape_feature:
+            return None
```
Comment (lucianopaz) [inline pytensor/tensor/random/rewriting/basic.py:278]: Should this really return `None`? If yes, add a `# pragma: no cover` to ignore the missing line coverage.

File: pytensor/tensor/random/rewriting/basic.py [hunk 5]
```diff
@@ -277,34 +269,35 @@ def is_nd_advanced_idx(idx, dtype):
...
+
+        # Use shape_feature to facilitate inferring final shape.
+        # Check that neither the RV nor the old Subtensor are in the shape graph.
+        output_shape = fgraph.shape_feature.shape_of.get(indexed_rv, None)
+        if output_shape is None or {indexed_rv, rv} & set(ancestors(output_shape)):
+            return None
```
Comment (lucianopaz) [inline pytensor/tensor/random/rewriting/basic.py:284]: Same as above.

**Review by lucianopaz — COMMENTED** [review]
I think this is Ok. But in all honesty, someone else should also have a look. I couldn't spend the necessary time to dig deep into the stuff that I didn't fully understand. I left a few questions though

File: pytensor/tensor/random/rewriting/basic.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/random/rewriting/basic.py:193]: A dictionary allows you to replace any keys by the values. It's an alternative to the default you're used to, which would be something like `dict(zip(node.outputs, new_outputs))`

File: pytensor/tensor/random/rewriting/basic.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/random/rewriting/basic.py:164]: I realized there's no need to introduce size if it was not defined in the original RV

File: pytensor/tensor/random/rewriting/basic.py (hunk 3 above)
Comment (ricardoV94) [reply pytensor/tensor/random/rewriting/basic.py:228]: I didn't remove but restricted it to the only case that's needed, which is when the RV had a size defined

File: pytensor/tensor/random/rewriting/basic.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/random/rewriting/basic.py:278]: Yes, it should return None, we need the feature for this rewrite case to work properly

File: pytensor/tensor/random/rewriting/basic.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/tensor/random/rewriting/basic.py:284]: Both were the behavior before so it's not a new thing

**Pull request merged**